core/attrlist.hpp

```diff
@@ -0,0 +1,177 @@
+    template <typename ObjT> friend class ObjList;
+
+  public:
+    BaseAttrList() = default;
+    virtual ~BaseAttrList() = default;
+
+    virtual size_t size() = 0;
+
+  protected:
+    virtual void resize(const size_t size) = 0;
+
+    virtual void reorder(std::vector<int>& order) = 0;
+
+    virtual void move(const size_t dest, const size_t src) = 0;
+
+    virtual void migrate(BinStream& bin, const size_t idx) = 0;
+
+    virtual void process_bin(BinStream& bin, const size_t idx) = 0;
+};
+
+template <typename ObjT, typename AttrT>
+class AttrList : public BaseAttrList {
+  public:
+    AttrList() = default;
+    ~AttrList() = default;
+
+    explicit AttrList(ObjListData<ObjT>* objlist_data_ptr) {
+        master_data_ptr_ = objlist_data_ptr;
+        data_.resize(master_data_ptr_->get_size());
+    }
+
+    // Size
+    inline size_t size() override { return data_.size(); }
+
+    // Get data
+    std::vector<AttrT>& get_data() { return data_; }
+
+    // Getter
+    // WARNING Attribute value after 'resize' is undefined
+    AttrT& get(const size_t idx) {
+        size_t objlist_size = master_data_ptr_->get_size();
+        ASSERT_MSG(idx < objlist_size, "AttrList.get: invalid index");
+        if (idx >= data_.size()) {
+            data_.resize(objlist_size);
+        }
+        return data_[idx];
+    }
+
+    AttrT& get(const ObjT& obj) {
+        size_t idx = master_data_ptr_->index_of(&obj);
+        return this->get(idx);
+    }
+
+    // Overload []
+    inline AttrT& operator[](const size_t idx) { return this->get(idx); }
+
+    // Setter
+    void set(const size_t idx, AttrT&& attr) {
+        size_t objlist_size = master_data_ptr_->get_size();
+        ASSERT_MSG(idx < objlist_size, "AttrList.set: invalid index");
+        if (idx >= data_.size()) {
+            data_.resize(objlist_size);
+        }
+        data_[idx] = std::move(attr);
+    }
+
+    void set(const size_t idx, const AttrT& attr) {
+        size_t objlist_size = master_data_ptr_->get_size();
+        ASSERT_MSG(idx < objlist_size, "AttrList.set: invalid index");
+        if (idx >= data_.size()) {
+            data_.resize(objlist_size);
+        }
+        data_[idx] = attr;
+    }
+
+    void set(const ObjT& obj, AttrT&& attr) {
+        size_t idx = master_data_ptr_->index_of(&obj);
+        this->set(idx, std::move(attr));
+    }
+
+    void set(const ObjT& obj, const AttrT& attr) {
+        size_t idx = master_data_ptr_->index_of(&obj);
+        this->set(idx, attr);
+    }
+
+  protected:
+    // Resize
+    inline void resize(const size_t size) override { data_.resize(size); }
+
+    // Reorder the list according to a permutaion
+    inline void reorder(std::vector<int>& order) override { reorder(data_, order); }
+
+    // Move j_th data to i_th
+    inline void move(const size_t dest, const size_t src) override {
+        data_[dest] = std::move(data_[src]);
+    }
+
+    void reorder(std::vector<AttrT>& data, std::vector<int> order) {
```

Comment:
@Yuzhen11 No, because ```reorder``` will modify ```order```, which is not desired as ```order``` might be applied on more than one list. So copying cannot be avoided in this case, I think.